### PR TITLE
[SYCL]: basic support of contexts with multiple devices in Level-Zero

### PR DESCRIPTION
**Pull request by smaslov-intel**
PI context in Level-Zero can now be created with multiple PI devices in it.
Compile/link/build support is still limited to single-device, though, and will be addressed in subsequent PR.

Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -3100,7 +3129,11 @@ pi_result piSamplerCreate(pi_context Context,
   assert(Context);
   assert(RetSampler);
 
-  ze_device_handle_t ZeDevice = Context->Device->ZeDevice;
+  // Have the "0" device in context to own the sampler. Rely on Level-Zero
+  // drivers to perform migration as necessary for sharing it across multiple
```
Comment (jandres742) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3133]: @smaslov-intel : What do you mean by rely on Level-Zero for migration across devices?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3133]: @jandres742 , I mean that the image/sampler created on one device with `zeImageCreate`/`zeSamplerCreate` are made available for access in other devices of this SYCL context without any explicit transfers in the Level-Zero plugin. Note that for buffers I use `zeMemAllocShared` but image/sampler don't have **shared** counterparts.

**Review by againull — APPROVED** [review]

**Review by againull — APPROVED** [review]
@smaslov-intel could you please resolve conflicts?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (jandres742) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3133]: L0 does not implicitly migrate between devices, nor for images nor buffers. From here, https://spec.oneapi.com/level-zero/latest/core/PROG.html#memory:

Shared allocations: Migratable to:  Another Device = Optional

Additionally. Access from one device to another device's buffers and images are not supported at all moments. Proper way of confirming such access exist is using `zeDeviceCanAccessPeer()` https://spec.oneapi.com/level-zero/latest/core/api.html?highlight=canaccess#zedevicecanaccesspeer

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3133]: Thanks @jandres742 !

> L0 does not implicitly migrate between devices, nor for images nor buffers.

Not even between sub-devices of the same root device?

> Shared allocations: Migratable to: Another Device = Optional

So what `Optional` really means? Is this an option of what? 

> not supported at all moments

At what circumstances can we rely on it?

> Proper way of confirming such access exist is using zeDeviceCanAccessPeer()

But we need a way to guarantee such sharing between devices in the same SYCL context.
Is there a way to force the migration that is optional?
Should we just copy explicitly if no?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3161]: @jandres742 , ping

I've added TODO comments for now and would like to proceed with this PR, please approve if you are OK with that.

Comment (jandres742) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3159]: > Not even between sub-devices of the same root device?

There's no need for migrations between sub-devices, as they are part of the same device and each sub-device can access another sub-device's memory. Issue is when we are are referring about migrations between two separate physical devices.

>So what Optional really means? Is this an option of what?

That means that it depends on HW and SW (L0 driver, and kernel driver) support. Not always would be present for all platforms.

>But we need a way to guarantee such sharing between devices in the same SYCL context.
>Is there a way to force the migration that is optional?
>Should we just copy explicitly if no?

Exactly. The only way of _enforcing_ migrations between devices is to implement a SW fallback with bounce buffering in the host.

Comment (jandres742) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3159]: Looks good 👍

**Review by againull — APPROVED** [review]

**Pull request merged**